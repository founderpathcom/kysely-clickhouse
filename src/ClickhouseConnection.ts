import {
  CompiledQuery,
  DatabaseConnection,
  QueryResult
} from 'kysely';

import { createClient } from '@clickhouse/client'
import { ClickhouseDialectConfig } from '.';
import { NodeClickHouseClient } from '@clickhouse/client/dist/client';
import { randomUUID } from 'node:crypto';

export class ClickhouseConnection implements DatabaseConnection {
  #client: NodeClickHouseClient

  constructor(config: ClickhouseDialectConfig) {
    this.#client = createClient({
      ...config.options,
      clickhouse_settings: {
        ...config.options?.clickhouse_settings,
        date_time_input_format: 'best_effort',
      },
      session_id: randomUUID(),
    })
  }

  prepareQuery<O>(compiledQuery: CompiledQuery): string {
    let i = 0
    const compiledSql = compiledQuery.sql.replace(/\?/g, () => {
      const param = compiledQuery.parameters[i++]

      if (typeof param === 'number') {
        return `${param}`
      }

      // should never happen
      if (typeof param !== 'string') {
        return `'${JSON.stringify(param)}'`
      }

      return `'${param.replace(/'/gm, `\\'`).replace(/\\"/g, '\\\\"')}'`
    })

    return compiledSql.replace(
      /^update ((`\w+`\.)*`\w+`) set/i,
      "alter table $1 update"
    )
  }

  async executeQuery<O>(compiledQuery: CompiledQuery): Promise<QueryResult<O>> {
    if (compiledQuery.query.kind === 'InsertQueryNode') {
      // @ts-expect-error: fix types
      if (!compiledQuery.query.values?.values) {
        // handle complex query (ex: INSERT INTO ... SELECT ...)
        const query = this.prepareQuery(compiledQuery)
        const result = await this.#client.query({
          query,
          clickhouse_settings: {
            date_time_input_format: 'best_effort',
          },
        });

        const summary = result.response_headers['x-clickhouse-summary']
        const summaryObject = JSON.parse(
          (Array.isArray(summary) ? summary[0] : summary) ?? '{}'
        )

        return {
          rows: [],
          numAffectedRows: BigInt(summaryObject.written_rows ?? 0),
          numChangedRows: BigInt(summaryObject.written_rows ?? 0),
        }
      }

      const values = [
        compiledQuery.query.columns?.map(c => c.column.name) ?? [],
        // @ts-expect-error: fix types
        ...compiledQuery.query.values?.values.map(v => v.values) ?? [],
      ]
      
      const schema = compiledQuery.query.into?.table?.schema?.name;
      const table = compiledQuery.query.into?.table.identifier.name ?? "";
      const fullQualifiedTable = schema ? `${schema}.${table}` : table
      const resultSet = await this.#client.insert({
        table: fullQualifiedTable,
        format: 'JSONCompactEachRowWithNames',
        values,
        clickhouse_settings: {
          date_time_input_format: 'best_effort',
        },
      })

      return {
        rows: [],
        numAffectedRows: BigInt(resultSet.summary?.written_rows ?? 0),
        numChangedRows: BigInt(resultSet.summary?.written_rows ?? 0),
      }
    }

    if (compiledQuery.query.kind === 'SelectQueryNode') {
      const query = this.prepareQuery(compiledQuery)

      const resultSet = await this.#client.query({
        query,
      })
      const response = await resultSet.json()

      return {
        rows: response.data as O[],
      }
    }

    await this.#client.command({
      query: this.prepareQuery(compiledQuery),
      clickhouse_settings: {
        wait_end_of_query: 1,
      },
    })

    return {
      rows: [],
    }

  }

  async beginTransaction() {
    throw new Error('Transactions are not supported.');
  }

  async commitTransaction() {
    throw new Error('Transactions are not supported.');
  }

  async rollbackTransaction() {
    throw new Error('Transactions are not supported.');
  }

  async *streamQuery<O>(compiledQuery: CompiledQuery, chunkSize: number): AsyncIterableIterator<QueryResult<O>> {
    const query = this.prepareQuery(compiledQuery)

    const resultSet = await this.#client.query({
      query,
      format: 'JSONEachRow',
    })

    const stream = resultSet.stream()

    for await (const row of stream) {
      yield {
        rows: [row as O],
      }
    }
  }
}