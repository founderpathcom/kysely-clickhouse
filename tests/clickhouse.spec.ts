import { expect, test } from 'vitest'

import {  ClickhouseDialect } from '../src'
import { Kysely } from 'kysely';
import { sql } from 'kysely';

const kysely = new Kysely<any>({
  dialect: new ClickhouseDialect({
    options: {
      url: 'http://localhost:8123',
      username: 'clickhouse',
      password: 'clickhouse',
      database: 'test',
    }
  })
})


// @todo: make test environment work
test('introspection', async () => {
  const tables = await kysely.introspection.getTables()

  expect(tables).toMatchSnapshot()
})

test('insert', async () => {
  const q = kysely.insertInto('company_metrics').values([{
    company_id: 1,
    date: '2025-04-15',
    mrr: 100,
  },
  {
    company_id: 2,
    date: '2025-04-15',
    mrr: 200,
    }]);
  
  const compiled = q.compile();
  expect(compiled).toMatchSnapshot()
  
  // await q.execute();
})

test('complicated insert', async () => {
  const q = kysely
  .insertInto('company_metrics')
  .columns(['company_id', 'date', 'mrr'])
  .expression(
    kysely.selectFrom('company_metrics').select([sql`3`.as('company_id'), 'date', sql`mrr + 300`.as('mrr')]),
  )

  const compiled = q.compile();

  expect(compiled).toMatchSnapshot()

  // await q.execute()
})