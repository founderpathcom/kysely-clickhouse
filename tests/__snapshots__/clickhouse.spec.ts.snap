// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`complicated insert 1`] = `
{
  "parameters": [],
  "query": {
    "columns": [
      {
        "column": {
          "kind": "IdentifierNode",
          "name": "company_id",
        },
        "kind": "ColumnNode",
      },
      {
        "column": {
          "kind": "IdentifierNode",
          "name": "date",
        },
        "kind": "ColumnNode",
      },
      {
        "column": {
          "kind": "IdentifierNode",
          "name": "mrr",
        },
        "kind": "ColumnNode",
      },
    ],
    "into": {
      "kind": "TableNode",
      "table": {
        "identifier": {
          "kind": "IdentifierNode",
          "name": "company_metrics",
        },
        "kind": "SchemableIdentifierNode",
      },
    },
    "kind": "InsertQueryNode",
    "replace": undefined,
    "values": {
      "from": {
        "froms": [
          {
            "kind": "TableNode",
            "table": {
              "identifier": {
                "kind": "IdentifierNode",
                "name": "company_metrics",
              },
              "kind": "SchemableIdentifierNode",
            },
          },
        ],
        "kind": "FromNode",
      },
      "kind": "SelectQueryNode",
      "selections": [
        {
          "kind": "SelectionNode",
          "selection": {
            "alias": {
              "kind": "IdentifierNode",
              "name": "company_id",
            },
            "kind": "AliasNode",
            "node": {
              "kind": "RawNode",
              "parameters": [],
              "sqlFragments": [
                "3",
              ],
            },
          },
        },
        {
          "kind": "SelectionNode",
          "selection": {
            "column": {
              "column": {
                "kind": "IdentifierNode",
                "name": "date",
              },
              "kind": "ColumnNode",
            },
            "kind": "ReferenceNode",
            "table": undefined,
          },
        },
        {
          "kind": "SelectionNode",
          "selection": {
            "alias": {
              "kind": "IdentifierNode",
              "name": "mrr",
            },
            "kind": "AliasNode",
            "node": {
              "kind": "RawNode",
              "parameters": [],
              "sqlFragments": [
                "mrr + 300",
              ],
            },
          },
        },
      ],
    },
  },
  "sql": "insert into \`company_metrics\` (\`company_id\`, \`date\`, \`mrr\`) select 3 as \`company_id\`, \`date\`, mrr + 300 as \`mrr\` from \`company_metrics\`",
}
`;

exports[`insert 1`] = `
{
  "parameters": [
    1,
    "2025-04-15",
    100,
    2,
    "2025-04-15",
    200,
  ],
  "query": {
    "columns": [
      {
        "column": {
          "kind": "IdentifierNode",
          "name": "company_id",
        },
        "kind": "ColumnNode",
      },
      {
        "column": {
          "kind": "IdentifierNode",
          "name": "date",
        },
        "kind": "ColumnNode",
      },
      {
        "column": {
          "kind": "IdentifierNode",
          "name": "mrr",
        },
        "kind": "ColumnNode",
      },
    ],
    "into": {
      "kind": "TableNode",
      "table": {
        "identifier": {
          "kind": "IdentifierNode",
          "name": "company_metrics",
        },
        "kind": "SchemableIdentifierNode",
      },
    },
    "kind": "InsertQueryNode",
    "replace": undefined,
    "values": {
      "kind": "ValuesNode",
      "values": [
        {
          "kind": "PrimitiveValueListNode",
          "values": [
            1,
            "2025-04-15",
            100,
          ],
        },
        {
          "kind": "PrimitiveValueListNode",
          "values": [
            2,
            "2025-04-15",
            200,
          ],
        },
      ],
    },
  },
  "sql": "insert into \`company_metrics\` (\`company_id\`, \`date\`, \`mrr\`) values (?, ?, ?), (?, ?, ?)",
}
`;

exports[`introspection 1`] = `
[
  {
    "columns": [
      {
        "comment": undefined,
        "dataType": "Int32",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "company_id",
      },
      {
        "comment": undefined,
        "dataType": "Date",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "date",
      },
      {
        "comment": undefined,
        "dataType": "Int32",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr",
      },
    ],
    "isView": false,
    "name": "company_metrics",
    "schema": "test",
  },
]
`;
